core/market/decentralized/src/protocol/discovery.rs

```diff
@@ -52,60 +58,81 @@ pub struct Discovery {
     inner: Arc<DiscoveryImpl>,
 }
 
+pub(super) struct ReceiveHandlers {
+    offer_ids: HandlerSlot<OfferIdsReceived>,
+    offers: HandlerSlot<OffersReceived>,
+}
+
 pub struct DiscoveryImpl {
-    offer_received: HandlerSlot<OfferReceived>,
+    identity: Arc<dyn IdentityApi>,
+
+    receive: Mutex<ReceiveHandlers>,
     offer_unsubscribed: HandlerSlot<OfferUnsubscribed>,
-    _retrieve_offers: HandlerSlot<RetrieveOffers>,
+    get_offers_request: HandlerSlot<GetOffers>,
 }
 
 impl Discovery {
     /// Broadcasts offer to other nodes in network. Connected nodes will
     /// get call to function bound in `offer_received`.
-    pub async fn broadcast_offer(&self, offer: ModelOffer) -> Result<(), DiscoveryError> {
-        log::info!("Broadcasting offer [{}].", &offer.id);
-
-        let original_sender = offer.node_id.clone();
-        let bcast_msg = SendBroadcastMessage::new(OfferReceived { offer });
+    pub async fn broadcast_offers(
+        &self,
+        offers: Vec<SubscriptionId>,
+    ) -> Result<(), DiscoveryError> {
+        let default_id = self.default_identity().await?;
+        let bcast_msg = SendBroadcastMessage::new(OfferIdsReceived { offers });
 
-        let _ = bus::service(net::local::BUS_ID)
-            .send_as(original_sender, bcast_msg) // TODO: should we send as our (default) identity?
+        // TODO: We shouldn't use send_as. Put identity inside broadcasted message instead.
+        let _ = bus::service(broadcast::BUS_ID)
+            .send_as(default_id, bcast_msg) // TODO: should we send as our (default) identity?
             .await?;
         Ok(())
     }
 
-    pub async fn broadcast_unsubscribe(
+    /// Ask remote Node for specified Offers.
+    pub async fn get_offers(
```

Comment:
to indicate it is remote operation, not local store select
```suggestion
    pub async fn get_offers_from(
```

Reply:
Hmmm I think it's obvious. You call these method on Discovery interface and pass NodeId in parameter. I don't think it can be missinterpreted

Reply:
It was not to me. 
I propose below naming scheme for this file:
* sender `broadcast_offers` and its receiver `on_broadcast_offers`
* `broadcast_unsubscribes` and `on_broadcast_unsubscribes`
* `retrieve_offers` --> `on_retrieve_offers`

`get_offers` resembles local db operation rather than remote

Reply:
And consequently `OffersReceived` rename to `OffersRetrieved` as it reflects this operation is triggered by us and not via broadcast.

Reply:
Fixed

Reply:
rethinked again and also fixed